### DISCUSSION TITLE
Add support for TSVs

### DISCUSSION THREAD

**Discussion by jessica-brooks** [bxparks/bigquery-schema-generator#91] [Ideas]
Adding in 
`
        elif self.input_format == 'tsv':
            reader = csv.DictReader(input_data, delimiter='\t')
`
around [line 179](https://github.com/bxparks/bigquery-schema-generator/blob/develop/bigquery_schema_generator/generate_schema.py#L179) adds this functionality. 
I've tested this locally and haven't encountered any issues (yet).

**Comment by bxparks** [comment]
I'm not opposed to adding more CSV/TSV features...

My concern is that the `DictReader` class supports a large number of options, and currently, the `SchemaGenerator` class does not allow the user to specify those options. I'm worried that I will get a never-ending slow drip of requests to add one or more of those options.

Maybe the `deduce_schema(self, input_file, *, schema_map)` needs to be generalized. I *think* we could add an `--input_format='reader'` option which accepts an instance of the `csv.DictReader` created by the calling code:

```python
if self.input_format == 'csv':
    reader = csv.DictReader(input_data)
elif self.input_format == 'reader':
    reader = input_data
...
elif self.input_format == 'dict':
    reader = input_data
...
```

Funny enough, due to Python's use of duck-typing (edited), `input_format='reader'` is exactly the same as `input_format='dict'`.

**Reply by bxparks** [reply]
And now that I'm more awake, the `input_format='dict'` is exactly what you want. Because `'dict'` actually means "iterable of dict", and `DictReader` is exactly that, it is an iterable of `dict` on top of the CSV or TSV file.
So we don't need an `input_format='reader'`. But I think the README.md can be improved to explain this.

**Reply by bxparks** [reply]
After some experimenting, it turned out that using `input_format='dict'` with `csv.DictReader` works, but there is a semantic difference between the two. The `'dict'` option assumes that the data can be a hierarchy, like `'json'` so the `infer_mode=True` flag is unable to determine that `mode=REQUIRED` is possible instead of `mode=NULLABLE`.

So I created a new `input_format='csvdictreader'` option, specially designed to be used with `csv.DictReader`. Check out the documentation at https://github.com/bxparks/bigquery-schema-generator#SchemaGeneratorDeduceSchemaFromCsvDictReader, and the example code at [examples/tsvreader.py](https://github.com/bxparks/bigquery-schema-generator/blob/develop/examples/tsvreader.py).

This code is available on the `develop` branch.
